Frameworks/Foundation/NSOperation.mm

```diff
@@ -111,32 +115,36 @@ - (id)init {
  @Status Interoperable
 */
 - (BOOL)isReady {
-    std::lock_guard<std::recursive_mutex> lock(_readyLock);
+    std::lock_guard<std::recursive_mutex> lock(_dependenciesLock);
     return _ready;
 }
 
 /**
  @Status Interoperable
 */
 - (void)addDependency:(id)operation {
-    std::lock_guard<std::recursive_mutex> lock(_dependenciesLock);
+    std::lock_guard<std::recursive_mutex> dependenciesLock(_dependenciesLock);
     [self willChangeValueForKey:@"dependencies"];
     [_dependencies addObject:operation];
-    [operation addObserver:self forKeyPath:@"isFinished" options:0 context:(void*)NSOperationContext];
+    [operation addObserver:self forKeyPath:@"isFinished" options:NSKeyValueObservingOptionInitial context:(void*)NSOperationContext];
     [self didChangeValueForKey:@"dependencies"];
+    _outstandingDependenciesCount++; // will be decremented if necessary in observeValueForKey, due to NSKeyValueObservingOptionInitial
```

Comment:
NSKeyValueObservingOptionInitial sends an immediate observation for the current state, before addObserver returns (and in the same callstack). Therefore, if the dependency is already finished, we will increment and decrement _outstandingDependenciesCount before addDependency returns, while holding the appropriate locks. According to documentation, "The same dependency should not be added more than once to the receiver, and the results of doing so are undefined."

---
In reply to: [72513161](https://github.com/Microsoft/WinObjC/pull/691#discussion_r72513161) [](ancestors = 72513161)